x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/CountingRunner.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * An utility class that keeps a counter when executing runnables.
+ * It is designed to help minimizing the possibility of caching stable results in a {@link Cache}.
+ */
+public class CountingRunner {
```

Comment:
As discussed, we both agree that encapsulating a cache inside this class is the wrong abstraction (which is `InvalidationCountingCacheWrapper`), although I struggle to put it in words in a comment.

I still tend to have more ambitious hopes for this cache invalidation pattern, but I very much fear we can sunk in it and not achieve much.

Overall, I think abstraction here is good enough!
But its rich history is showing in the description and in the names. We should polish these.

I would think about this class as a special type of atomic integer, which has a fancy method "run if integer constant", such that the run block of code cannot overlap with changes (increments) to the integer.
It's hard to explain it's usage in the caching context, so I would steer away from it completely.
For example, we rely that changes to the underlying data (eg api key docs) are performed **before** calling `increment` and that we call `getCount` **before** we retrieve the underlying data, so that we ensure the sequence `modify data` -> `increment` -> `getCount` -> `read data` -> `cache data` .

My explicit _suggestion_ is to rename to `CounterCompare`, or anythign else that inspires that the runnable and the counter cannot overlap.

Reply:
Another suggestion, if you think this is unlikely to be used outside of caching contexts, it might be useful to make it a static inner class of `Cache`.

Reply:
After some thoughts and taking your advice onboard, I came up with the name `LockingAtomicCounter`. It is an unexciting name, but I think it conveys the major idea here: a counter with RWLock to ensure its value stays the same until the reading side finishes its task.

I think your analogy of `AtomicInteger` is a great one. In concept, `compareAndSet` and `compareAndRun` are very similiar. They are both executing some operation if the counter value matches the expectation. The only difference is that the operation of the former is a single setValue, but the later can have an arbitrary Runnable. This is also the reason that I went with the name of `LockingAtomicCounter`, which I hope makes its behaviour easier to understand.

I updated the comments accordingly. 

As for the location of this class, I'd prefer to have it stay inside xpack security for the time being because:
1. The usages that I can think of for now are security caches
2. It in theory operates at a higher level than a single cache, e.g. `ApiKeyDocCache` is a wrapper for two vanila caches.
3. The way it ends right now really does not say anything about caching anymore. So I have hope that it could still be useful elsewhere.

Reply:
> I came up with the name LockingAtomicCounter

SGTM